KOGITO-7265: [KSW-Guides] Authentication support for OpenAPI services (#9)
kiegroup/kogito-docs: serverlessworkflow/modules/ROOT/pages/security/authention-support-for-openapi-services.adoc

> = Authentication support for OpenAPI services

Author: Not sure if it's a good idea, the generated result looks really bad. When we work with antora it looks like that's correct way to set the page titles is this that we have.

> = Authentication support for OpenAPI services

Reviewer: what do you mean, we've adding it so far and we haven't found any issue:
e.g. https://github.com/kiegroup/kogito-docs/pull/14/files#diff-97d7cbd59dfe85ab12496fe70980a886c25f3ad4cee435d1f50cce3aee9e12c1R1

> = Authentication support for OpenAPI services

Author: The link you passed before had this "==============" , I thought it was a fixed string and did a quick run, and the generated page looked like this, that looked horrible.
![Screenshot from 2022-06-10 16-14-41](https://user-images.githubusercontent.com/2431454/173085152-95741f58-cf68-47fe-b349-c87d7b3123c3.png)

I didn't realize that the number of "="s was important.
Now it looks well.
Fixed

      }
    }
----

In general, to configure the credentials for accessing the secured operations, and other security parameters, means to configure application properties related to those security schemes.
For that, we will use configuration keys that are composed using this pattern:
[source, properties]
----
quarkus.openapi-generator.[filename].[security_scheme_name].auth.[auth_property_name]
----
where:

- `filename` is the sanitized name of the file containing the OpenAPI specification, for example `security_example_json`.
- `security_scheme_name` is the sanitized name of the security scheme object definition in the OpenAPI file, for example `http_basic_example` or `api_key_example`.
- `auth_property_name` is name of the property to configure and depends on the security scheme type, for example `username`.

[NOTE#sanitize_note]
====
To sanitize these names we follow the https://github.com/eclipse/microprofile-config/blob/master/spec/src/main/asciidoc/configsources.asciidoc#environment-variables-mapping-rules[Environment Variables Mapping Rules] from Microprofile Configuration. Any non-alphabetic characters are replaced by an underscore `_`.
====

[IMPORTANT]
====

> The security schemes defined in an OpenAPI specification are global to all the operations on that file, which means that the configurations made for a particular security scheme apply to all the operations secured with it.

Reviewer: ```suggestion
The security schemes defined in an OpenAPI specification are global to all the operations on that file, which means that the configurations made for a particular security scheme apply the other secured operations as well..
```

 +
Enables the token propagation for all the operations that are secured with the `oauth-example` scheme in the `security-example.json` file. By default, the standard HTTP Authorization header is propagated.

| `quarkus.openapi-generator.[filename].auth.[security_scheme_name].header-name=[http_header_name]`
| `quarkus.openapi-generator.security_example_json.auth.oauth_example.header-name=MyHeaderName` +
 +
Stays that the authorization token to propagate will be read from the HTTP header `MyHeaderName` instead of the standard HTTP `Authorization` header.

|===

[IMPORTANT]
====
The authorization tokens are propagated as long the Serverless Workflow doesn't reach a "waiting" state, for example to wait for an event to arrive. When the Serverless Workflow is resumed the tokens will be no longer propagated.
====

== Going further

The Serverless Workflows OpenAPI support is based on the https://github.com/quarkiverse/quarkus-openapi-generator[Quarkus OpenAPI Generator] extension. While not mandatory, it could be an interesting reading if you want to know more details or even other usages not related to the Serverless Workflows.

> 

Reviewer: is there a room for what's next here?

Reviewer 2: Not sure. This is the real end-road. People here already know about the services orchestration using OpenAPI. I think it's fine as it is unless @wmedvede has another idea.

Author: hahaha, this is sad.
I've just added the what's next section

By using a simple naming convention, the token management operations will be delegated to a Quarkus `OidcClient`.

For the example above, the token management operations for accessing the `sayHelloOauth2` operation will be delegated to the OidcClient `oauth_example`.

You can configure this `OidcClient` as:

[source, properties]
----
quarkus.oidc-client.oauth_example.auth-server-url=https://example.com/oauth
quarkus.oidc-client.oauth_example.token-path=/tokens
quarkus.oidc-client.oauth_example.discovery-enabled=false
quarkus.oidc-client.oauth_example.client-id=kogito-app
quarkus.oidc-client.oauth_example.grant.type=client
quarkus.oidc-client.oauth_example.credentials.client-secret.method=basic
quarkus.oidc-client.oauth_example.credentials.client-secret.value=secret
----

The configuration suffix `quarkus.oidc-client.oauth_example` is exclusive for the schema defined in the specification file and the scheme name is sanitized by applying the rules described <<sanitize_note, above>>.

For this to work you **must** add the https://quarkus.io/guides/security-openid-connect-client#oidc-client-filter[Quarkus OIDC Client Filter Extension] to your project:

[source, xml]

> ----

Author: A tip was added here to refresh the reader memory and help him to use the Quarkus CLI too.
![Screenshot from 2022-06-14 15-42-29](https://user-images.githubusercontent.com/2431454/173592341-39c31803-dc5a-4f65-94ba-23ad1493f5b1.png)

> ----

Author: well, it depends...., I believe this guide is more a kind of documentation guide rather than a step-by-step.one. A reader that reach this guide has normally already went through step-by-step guides, like the "create a hello world SW", and is a developer that will directly copy paste the dependency. I think It's ok.

 +
Enables the token propagation for all the operations that are secured with the `oauth-example` scheme in the `security-example.json` file. By default, the standard HTTP Authorization header is propagated.

| `quarkus.openapi-generator.[filename].auth.[security_scheme_name].header-name=[http_header_name]`
| `quarkus.openapi-generator.security_example_json.auth.oauth_example.header-name=MyHeaderName` +
 +
Stays that the authorization token to propagate will be read from the HTTP header `MyHeaderName` instead of the standard HTTP `Authorization` header.

|===

[IMPORTANT]
====
The authorization tokens are propagated as long the Serverless Workflow doesn't reach a "waiting" state, for example to wait for an event to arrive. When the Serverless Workflow is resumed the tokens will be no longer propagated.
====

== Going further

The Serverless Workflows OpenAPI support is based on the https://github.com/quarkiverse/quarkus-openapi-generator[Quarkus OpenAPI Generator] extension. While not mandatory, it could be an interesting reading if you want to know more details or even other usages not related to the Serverless Workflows.

> == What's next?

Reviewer 3: Agreed, we don't necessarily need to describe all options in detail here.  But having a hint so people can search in the right direction or a reference to more details would imho be useful.  @spolti you might be able to help here?

> == What's next?

Reviewer 2: Also, I think we will have a full configuration example in the next guide, orchestrating 3rd services...

.security-example.json
[source, json]
----
{
  "openapi": "3.1.0",
  "info": {
    "title": "Api Key Scheme Example",
    "version": "1.0"
  },
  "paths": {
    "/hello-with-api-key": {
      "get": {
        "operationId": "sayHelloApiKey",
        "responses": {
          "200": {
            "description": "OK",
            "content": {
              "text/plain": {
                "schema": {
                  "type": "string"
                }
              }
            }
          }
        },
        "security": [{"api-key-example" : []}]
      }
    }
  },
  "components": {
    "securitySchemes": {
      "api-key-example": {
        "type": "apiKey",
        "name": "api-key-name",
        "in": "header"
      }
    }
  }
}
----

**These are the supported configurations:**

[frame=all]
|===
|Description |Property Key |Example

|API Key
|`quarkus.openapi-generator.[filename].auth.[security_scheme_name].api-key`
|`quarkus.openapi-generator.security_example_json.auth.api_key_example.api-key=MY_KEY`

|===

> The API Key scheme has an additional property `name` that specifies where to add the API Key in the request token: _header_, _cookie_ or _query_.

Author: was rephrased better

    "version": "1.0"
  },
  "paths": {
    "/hello-with-oauth2": {
      "get": {
        "operationId": "sayHelloOauth2",
        "responses": {
          "200": {
            "description": "OK",
            "content": {
              "text/plain": {
                "schema": {
                  "type": "string"
                }
              }
            }
          }
        },
        "security": [{"oauth-example" : []}]
      }
    }
  },
  "components": {
    "securitySchemes": {
      "oauth-example": {
        "type": "oauth2",
        "flows": {
          "clientCredentials": {
            "authorizationUrl": "https://example.com/oauth",
            "tokenUrl": "https://example.com/oauth/token",
            "scopes": {
            }
          }
        }
      }
    }
  }
}
----

**These are the supported configurations:**

Unlike the `http basic`, `http bearer` and `apiKey` security schemes, the Oauth2 authentication relies on the https://quarkus.io/guides/security-openid-connect-client[Quarkus OpenId Connect and Oauth2 Clients and Filters].

By using a simple naming convention, the token management operations will be delegated to a Quarkus `OidcClient`.

> For the example above, the token management operations for accessing the `sayHelloOauth2` operation will be delegated to the OidcClient `oauth_example`.

Reviewer 4: Reading further, I would maybe merge another two following paragraphs as bullet points under the proposed list above, paragraphs until `For this to work you **must**...`

> For the example above, the token management operations for accessing the `sayHelloOauth2` operation will be delegated to the OidcClient `oauth_example`.

Author: this was already rephrased by @hmanwani-rh , I think it loos good now.

Following the <<_oauth2_authentication>> example presented above, these are the supported configurations:

[frame=all]
|===
|Property Key |Example

| `quarkus.openapi-generator.[filename].auth.[security_scheme_name].token-propagation=[true,false]`
| `quarkus.openapi-generator.security_example_json.auth.oauth_example.token-propagation=true` +
 +
Enables the token propagation for all the operations that are secured with the `oauth-example` scheme in the `security-example.json` file. By default, the standard HTTP Authorization header is propagated.

| `quarkus.openapi-generator.[filename].auth.[security_scheme_name].header-name=[http_header_name]`
| `quarkus.openapi-generator.security_example_json.auth.oauth_example.header-name=MyHeaderName` +
 +
States that the authorization token to propagate will be read from the HTTP header `MyHeaderName` instead of the standard HTTP `Authorization` header.

|===

[IMPORTANT]
====
The authorization tokens are propagated as long the Serverless Workflow doesn't reach a "waiting" state, for example to wait for an event to arrive. When the Serverless Workflow is resumed the tokens will be no longer propagated.
====

== What's next?

> This guide covered the authentication types that are supported to access the OpenAPI service operations and the respective configurations.

Reviewer: It can probably be removed.

Reviewer: I mean, this: "This guide covered the authentication types that are supported to access the OpenAPI service operations and the respective configurations."

can be removed.

Author: ah, yes, now I understand. Yeah, it's again a styling criteria, many times when you write an article, blog, paper, etc, you finish with a short "summary", "phrase", etc, and then, you write the conclusions, next steps, etc.

We recommend continuing the journey with the {orchestrating-third-party-services-with-oauth2}[Orchestrating third party services with OAuth2 authentication] guide.

== Going further

> The Serverless Workflows OpenAPI support is based on the https://github.com/quarkiverse/quarkus-openapi-generator[Quarkus OpenAPI Generator] extension. While not mandatory, it could be an interesting reading if you want to know more details or even other usages not related to the Serverless Workflows.

Reviewer: I think this should be somewhere at the getting started guides.
@ricardozanini @krisv wdyt?

Reviewer 2: I think here's it's fine since users will get here only if they are interested in OpenAPI calls.
